File: src/main/java/com/campspot/dao/PunchDAO.java
package com.campspot.dao;

import com.campspot.api.Punch;
import org.joda.time.DateTime;
import org.skife.jdbi.v2.sqlobject.GetGeneratedKeys;
import org.skife.jdbi.v2.sqlobject.SqlQuery;
import org.skife.jdbi.v2.sqlobject.SqlUpdate;

import java.util.List;

public interface PunchDAO {
  @SqlUpdate("INSERT INTO Punch (start, end, category, description) VALUES (:start, :end, :category, :description)")
  @GetGeneratedKeys
  Long create(DateTime start, DateTime end, String category, String description);

  @SqlUpdate("UPDATE Punch SET start = :start, end = :end, category = :category, description = :description WHERE id = :id")
  void update(Long id, DateTime start, DateTime end, String category, String description);

  @SqlQuery("SELECT * FROM Punch WHERE start >= :start and end <= :end")
  List<Punch> listForDates(DateTime start, DateTime end);

  @SqlQuery("SELECT COUNT(id) > 0 FROM Punch WHERE start >= :start and end <= :end")
  Boolean anyInRange(DateTime start, DateTime end);

  @SqlQuery("SELECT * FROM Punch WHERE id = :id")
  Punch findById(Long id);
}

File: src/main/kotlin/com/campspot/dao/PunchDAO.kt
package com.campspot.dao

import com.campspot.api.Punch
import org.jdbi.v3.sqlobject.statement.GetGeneratedKeys
import org.jdbi.v3.sqlobject.statement.SqlQuery
import org.jdbi.v3.sqlobject.statement.SqlUpdate
import java.time.ZonedDateTime

interface PunchDAO: DAO {
  @SqlUpdate("INSERT INTO Punch (start, end, category, description) VALUES (:punch.start, :punch.end, :punch.category, :punch.description)")
  @GetGeneratedKeys
  fun create(punch: Punch): Long

  @SqlUpdate("UPDATE Punch SET start = :punch.start, end = :punch.end, category = :punch.category, description = :punch.description WHERE id = :punch.id")
  fun update(punch: Punch)

  @SqlQuery("SELECT * FROM Punch WHERE start < :end AND end > :start")
  fun findAllInDateRange(start: ZonedDateTime, end: ZonedDateTime): List<Punch>

  @SqlQuery("SELECT COUNT(id) > 0 FROM Punch WHERE start < :end AND end > :start")
  fun anyInRange(start: ZonedDateTime, end: ZonedDateTime): Boolean

  @SqlQuery("SELECT * FROM Punch WHERE id = :id")
  fun findById(id: Long): Punch

  @SqlQuery("SELECT * FROM Punch WHERE category = :category LIMIT 1")
  fun findFirstForCategory(category: String): Punch
}
